### PR TITLE
non_local_means : patch size argument for local mean and variance

### PR DESCRIPTION
**Pull request by riddhishb**
Added the argument which allows user to change the patch size used for computation of local mean and local variance in the non_local_means.
Addresses the issue brought up in #1178

### CODE REVIEW

File: dipy/denoise/nlmeans_block.pyx [hunk 1]
```diff
@@ -277,25 +278,25 @@ cdef double _local_mean(double[:, :, :]ima, int x, int y, int z) nogil:
...
 
-    return ss / 27.0
+    return ss / vol
 
 
 @cython.boundscheck(False)
 @cython.wraparound(False)
 @cython.cdivision(True)
-cdef double _local_variance(double[:, :, :] ima, double mean, int x, int y, int z) nogil:
+cdef double _local_variance(double[:, :, :] ima, double mean, int x, int y, int z, int p) nogil:
```
Comment (arokem) [inline dipy/denoise/nlmeans_block.pyx]: This is going over 80 characters. Please break into two lines, by adding a line-break somewhere.

File: dipy/denoise/nlmeans_block.pyx [hunk 2]
```diff
@@ -277,25 +278,25 @@ cdef double _local_mean(double[:, :, :]ima, int x, int y, int z) nogil:
...
     cdef int dx, dy, dz, nx, ny, nz
-    for px in range(x - 1, x + 2):
-        for py in range(y - 1, y + 2):
-            for pz in range(z - 1, z + 2):
+    for px in range(x - p, x + p + 1):
+        for py in range(y - p, y + p + 1):
+            for pz in range(z - p, z + p + 1):
                 if ((px >= 0 and py >= 0 and pz > 0) and
                         (px < dims0 and py < dims1 and pz < dims2)):
                     ss += (ima[px, py, pz] - mean) * (ima[px, py, pz] - mean)
```
Comment (arokem) [inline dipy/denoise/nlmeans_block.pyx]: Doesn't variance have to be normalized by the volume?

File: dipy/denoise/nlmeans_block.pyx (hunk 1 above)
Comment (riddhishb) [inline dipy/denoise/nlmeans_block.pyx]: cool!

File: dipy/denoise/nlmeans_block.pyx (hunk 2 above)
Comment (riddhishb) [inline dipy/denoise/nlmeans_block.pyx]: Variance is normalized by the count variable which is essentially the volume, just a gimmick of the code

**Comment by riddhishb** [comment]
@arokem Thats a good idea, let me add a test

File: dipy/denoise/nlmeans_block.pyx [hunk 3]
```diff
@@ -346,7 +349,8 @@ cpdef upfir(double[:, :] image, double[:] h):
 @cython.boundscheck(False)
 @cython.wraparound(False)
 @cython.cdivision(True)
-def nlmeans_block(double[:, :, :]image, double[:, :, :] mask, int patch_radius, int block_radius, double h, int rician):
+def nlmeans_block(double[:, :, :]image, double[:, :, :] mask, int patch_radius,
```
Comment (Garyfallidis) [inline dipy/denoise/nlmeans_block.pyx]: indentation issue

**Comment by ssheybani** [comment]
I'm comparing your nlmeans() and non_local_means(). But I'm not sure how we can use the sigma given by estimate_sigma function in noise_estimate.py to test your non_local_means(). Because non_local_means() expects a single float value for sigma, while estimate_sigma() gives an array (with as many elements as the gradient directions). 
What I've done so far is that I gave both nlmeans and non_local_means the average of the sigmas. Here's the result:
/home/saber/Courses/Semester_1/Garyfallidis_Lab/Denoising/NLMeans/Wu/nlm olv vs new.png

I also tried sigma=5 for both. Here is the result:
/home/saber/Courses/Semester_1/Garyfallidis_Lab/Denoising/NLMeans/Wu/nlm olv vs new sigma5.png

In both cases, it looks like non_local_means is oversmoothing.
By the way, for the execution time, what I observed was that for small datasets (smal portions of complete datasets), non_local_means is faster. but when the dataset is large enough, non_local_means is twice slower than nlmeans.

**Comment by samuelstjean** [comment]
While we are there, the other half (and most important I'd say) part of the issue is that the outer loop does iterate over j,i,k, but it passes to the inner function the indexes in the i,j,k order, so that effectively swap in your local block the x and y coordinate. 

Since a small region is mostly homogeneous, I guess that's why you don't see the problem much, but it could explain the problem reported above as the weights will be inflated, and less stuff gets compared because of that.

**Comment by arokem** [comment]
Closing. @skoudoro is going to follow up on this. We are looking to streamline things quite a bit before our 1.0.

**Pull request closed without merging**